pkg/generator/statemanager/statemanager.go

```diff
@@ -0,0 +1,227 @@
+
+	"github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+	"github.com/external-secrets/external-secrets/apis/generators/v1alpha1"
+	"github.com/external-secrets/external-secrets/pkg/generator/gc"
+)
+
+// Manager takes care of maintaining the state of the generators.
+// It provides the ability to commit and rollback the state of the generators,
+// which is needed when we have multiple generators that need to be created or
+// other operations which can fail.
+// The manager shall be used to modify the state of the generators on a given resource.
+// The user can choose any key to store the state of the generator on the "latest" field.
+// When state is moved to GC, manager will create a hash of the key and the generator state
+// and store it in the "GC" field.
+type Manager struct {
+	resource      v1beta1.GeneratorStateManagingResource
+	internalState []QueueItem
+}
+
+type QueueItem struct {
+	Rollback func() error
+	Commit   func() error
+}
+
+func New(resource v1beta1.GeneratorStateManagingResource) *Manager {
+	return &Manager{
+		resource: resource,
+	}
+}
+
+// Rollback will rollback the state of the generators.
+func (m *Manager) Rollback() error {
+	var errs []error
+	for _, item := range m.internalState {
+		if err := item.Rollback(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
+}
+
+// Commit will commit the state of the generators.
+func (m *Manager) Commit() error {
+	var errs []error
+	for _, item := range m.internalState {
+		if err := item.Commit(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
+}
+
+// FlagLatestStateForGC will flag the latest state for garbage collection after Commit.
+// It will be cleaned up later by the garbage collector.
+func (m *Manager) FlagLatestStateForGC(stateKey string) {
+	m.internalState = append(m.internalState, QueueItem{
+		Commit: func() error {
+			genState := m.resource.GetGeneratorState()
+			if genState.Latest == nil {
+				return nil
+			}
+			latest, ok := genState.Latest[stateKey]
+			if !ok {
+				return nil
+			}
+			gen, err := v1alpha1.GetGenerator(latest.Resource)
+			if err != nil {
+				return err
+			}
+			return m.ImmediateMoveStateToGC(latest.Resource, stateKey, gen, latest.State)
+		},
+	})
+}
+
+// MoveStateToGC will move the generator state to GC.
+// It will be cleaned up later by the garbage collector.
+func (m *Manager) MoveStateToGC(resource *apiextensions.JSON, stateKey string, gen v1alpha1.Generator, state v1alpha1.GeneratorProviderState) {
+	m.internalState = append(m.internalState, QueueItem{
+		Commit: func() error {
+			return m.ImmediateMoveStateToGC(resource, stateKey, gen, state)
+		},
+	})
+}
+
+// ImmediateMoveStateToGC will move the generator state to GC immediately.
+func (m *Manager) ImmediateMoveStateToGC(resource *apiextensions.JSON, stateKey string, gen v1alpha1.Generator, state v1alpha1.GeneratorProviderState) error {
```

Comment:
These two are a bit confusing API and usage wise. Like the first one sets `Commit` function the second one actually performs and Enqueue but at the look of the API I wouldn't have thought that.

Like, `Move` doesn't actually do anything until you call `Commit`. I would call this a setup or a set commit or something like that to make apparent that you'd have to do something actually afterwards.

WDYT?

Reply:
I agree, that's bad design/naming. Let me rethink that.